### PR TITLE
[plugin-ssh] Add step to configure SSH connections dynamically

### PR DESCRIPTION
**Pull request by valfirst**

### CODE REVIEW

File: docs/modules/plugins/pages/plugin-ssh.adoc [hunk 1]
```diff
@@ -16,80 +16,148 @@ It's allowed to configure unlimited number of SSH connections via mechanism of t
...
 
-NOTE: The properties marked with *bold* are mandatory.
+:connection-parameters-prefix: ssh.server.<key>.
+:parameters-name: properties
+include::partial$ssh-connection-parameters.adoc[]
 
-[cols="3,1,1,3", options="header"]
+== Steps
+
+=== Configure dynamic SSH conenction
```
Comment (ikalinin1) [inline docs/modules/plugins/pages/plugin-ssh.adoc:27]:
```suggestion
=== Configure dynamic SSH connection
```

File: docs/modules/plugins/pages/plugin-ssh.adoc [hunk 2]
```diff
@@ -16,80 +16,148 @@ It's allowed to configure unlimited number of SSH connections via mechanism of t
...
+----
+
+* `$connectionKey` - The key to assign to the creating SSH connection. In case if the key conflicts with a global connection key, the dynamic connection will take precedence within the story.
+* `$connectionParameters` - `ExamplesTable` with SSH connection parameters.
++
+:connection-parameters-prefix:
+:parameters-name: parameters
+include::partial$ssh-connection-parameters.adoc[]
+
+.Configure SSH conenction dynamically
```
Comment (ikalinin1) [inline docs/modules/plugins/pages/plugin-ssh.adoc:43]:
```suggestion
.Configure SSH connection dynamically
```

File: docs/modules/plugins/partials/ssh-connection-parameters.adoc [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+|`{connection-parameters-prefix}channel-type`
+a|`exec` +
+`shell`
+|`exec`
+a|In common *nix OpenSSH server:
+
+* The `shell` channel executes a login shell (as if you login with SSH terminal client).
+* The `exec` command takes a command as an "argument" and executes it in an isolated environment &ndash; still via user's default shell, but not as a "login" shell, what may cause significant differences in the command execution.
+
+With https://stackoverflow.com/a/56713547/2067574[less common SSH servers], the difference can be even more significant. Some servers may even not support one of the channels. It is also quite common that they seemingly support both, but one of them (typically the `exec`) is completely broken.
```
Comment (ikalinin1) [inline docs/modules/plugins/partials/ssh-connection-parameters.adoc:73]:
```suggestion
With https://stackoverflow.com/a/56713547/2067574[less common SSH servers], the difference can be even more significant. Some servers may not even support one of the channels. It is also quite common that they seemingly support both, but one of them (typically the `exec`) is completely broken.
```

File: vividus-plugin-ssh/src/main/resources/spring.xml [hunk 4]
```diff
@@ -11,15 +11,17 @@
 
     <context:component-scan base-package="org.vividus.ssh" />
 
-    <bean class="org.vividus.context.SshTestContext" />
+    <bean class="org.vividus.ssh.converter.CommandsConverter" />
```
Comment (ikalinin1) [inline vividus-plugin-ssh/src/main/resources/spring.xml:14]: What is the reason for this change?

File: vividus-plugin-ssh/src/test/java/org/vividus/ssh/JSchExecutorTests.java [hunk 5]
```diff
@@ -53,7 +53,7 @@
     @Test
     void shouldExecuteSuccessfullyWithoutAgentForwarding() throws JSchException, CommandExecutionException
     {
-        ServerConfiguration server = getDefaultServerConfiguration();
+        SshConnectionParameters server = getDefaultServerConfiguration();
```
Comment (ikalinin1) [inline vividus-plugin-ssh/src/test/java/org/vividus/ssh/JSchExecutorTests.java:56]:
```suggestion
        var server = getDefaultServerConfiguration();
```

File: vividus-plugin-ssh/src/test/java/org/vividus/ssh/JSchExecutorTests.java [hunk 6]
```diff
@@ -75,7 +75,7 @@ void shouldExecuteSuccessfullyWithoutAgentForwarding() throws JSchException, Com
     void shouldExecuteSuccessfullyWithAgentForwarding()
             throws AgentProxyException, JSchException, CommandExecutionException
     {
-        ServerConfiguration server = getDefaultServerConfiguration();
+        SshConnectionParameters server = getDefaultServerConfiguration();
```
Comment (ikalinin1) [inline vividus-plugin-ssh/src/test/java/org/vividus/ssh/JSchExecutorTests.java:78]:
```suggestion
        var server = getDefaultServerConfiguration();
```

File: vividus-plugin-ssh/src/test/java/org/vividus/ssh/JSchExecutorTests.java [hunk 7]
```diff
@@ -105,7 +105,7 @@ void shouldExecuteSuccessfullyWithAgentForwarding()
     @Test
     void shouldFailOnCommandExecutionError() throws JSchException
     {
-        ServerConfiguration server = getDefaultServerConfiguration();
+        SshConnectionParameters server = getDefaultServerConfiguration();
```
Comment (ikalinin1) [inline vividus-plugin-ssh/src/test/java/org/vividus/ssh/JSchExecutorTests.java:108]:
```suggestion
        var server = getDefaultServerConfiguration();
```

File: vividus-plugin-ssh/src/test/java/org/vividus/ssh/JSchExecutorTests.java [hunk 8]
```diff
@@ -134,7 +134,7 @@ protected SshOutput executeCommand(ServerConfiguration serverConfig, Commands co
     @Test
     void shouldFailOnChannelOpeningError() throws JSchException
     {
-        ServerConfiguration server = getDefaultServerConfiguration();
+        SshConnectionParameters server = getDefaultServerConfiguration();
```
Comment (ikalinin1) [inline vividus-plugin-ssh/src/test/java/org/vividus/ssh/JSchExecutorTests.java:137]:
```suggestion
        var server = getDefaultServerConfiguration();
```

File: vividus-plugin-ssh/src/test/java/org/vividus/ssh/JSchExecutorTests.java [hunk 9]
```diff
@@ -154,10 +154,11 @@ void shouldFailOnChannelOpeningError() throws JSchException
         }
     }
 
+    @SuppressWarnings("try")
     @Test
     void shouldFailOnJSchConfigurationError()
     {
-        ServerConfiguration server = getDefaultServerConfiguration();
+        SshConnectionParameters server = getDefaultServerConfiguration();
```
Comment (ikalinin1) [inline vividus-plugin-ssh/src/test/java/org/vividus/ssh/JSchExecutorTests.java:161]:
```suggestion
        var server = getDefaultServerConfiguration();
```

File: vividus-plugin-ssh/src/test/java/org/vividus/ssh/JSchExecutorTests.java [hunk 10]
```diff
@@ -171,18 +172,18 @@ void shouldFailOnJSchConfigurationError()
...
-        serverConfiguration.setAgentForwarding(false);
-        serverConfiguration.setPassphrase("passphrase");
-        serverConfiguration.setPrivateKey("privatekey");
-        serverConfiguration.setPublicKey("publickey");
-        serverConfiguration.setUsername("username");
-        serverConfiguration.setHost("host");
-        serverConfiguration.setPort(22);
-        serverConfiguration.setPassword("password");
-        return serverConfiguration;
+        SshConnectionParameters sshConnectionParameters = new SshConnectionParameters();
```
Comment (ikalinin1) [inline vividus-plugin-ssh/src/test/java/org/vividus/ssh/JSchExecutorTests.java:177]:
```suggestion
        var sshConnectionParameters = new SshConnectionParameters();
```

File: vividus-plugin-ssh/src/test/java/org/vividus/ssh/exec/SshExecExecutorTests.java [hunk 11]
```diff
@@ -40,10 +40,10 @@ void shouldReturnExecChannelType()
     void shouldConfigureChannel()
     {
         ChannelExec channel = mock(ChannelExec.class);
-        ServerConfiguration serverConfiguration = new ServerConfiguration();
-        serverConfiguration.setAgentForwarding(true);
-        serverConfiguration.setPseudoTerminalEnabled(true);
-        sshExecExecutor.configureChannel(channel, serverConfiguration);
+        SshConnectionParameters sshConnectionParameters = new SshConnectionParameters();
```
Comment (ikalinin1) [inline vividus-plugin-ssh/src/test/java/org/vividus/ssh/exec/SshExecExecutorTests.java:43]:
```suggestion
        var sshConnectionParameters = new SshConnectionParameters();
```

File: vividus-plugin-ssh/src/test/java/org/vividus/ssh/exec/SshExecutionManagerTests.java [hunk 12]
```diff
@@ -49,14 +49,14 @@ void shouldRunExecution(String channelType, Class<SshExecutor<? extends Channel>
     {
         SshExecutor<? extends Channel> executor = mock(executorClass);
         when(executor.getChannelType()).thenCallRealMethod();
-        ServerConfiguration serverConfiguration = new ServerConfiguration();
-        serverConfiguration.setChannelType(Optional.ofNullable(channelType));
+        SshConnectionParameters sshConnectionParameters = new SshConnectionParameters();
```
Comment (ikalinin1) [inline vividus-plugin-ssh/src/test/java/org/vividus/ssh/exec/SshExecutionManagerTests.java:52]:
```suggestion
        var sshConnectionParameters = new SshConnectionParameters();
```

File: vividus-plugin-ssh/src/test/java/org/vividus/ssh/exec/SshExecutorTests.java [hunk 13]
```diff
@@ -75,9 +75,9 @@ void testExecuteCommandsViaSshSuccessfully() throws Exception
         }).thenReturn(Boolean.TRUE);
         int exitStatus = 1;
         when(channel.getExitStatus()).thenReturn(exitStatus);
-        ServerConfiguration serverConfiguration = new ServerConfiguration();
+        SshConnectionParameters sshConnectionParameters = new SshConnectionParameters();
```
Comment (ikalinin1) [inline vividus-plugin-ssh/src/test/java/org/vividus/ssh/exec/SshExecutorTests.java:78]: va
```suggestion
        var sshConnectionParameters = new SshConnectionParameters();
```

File: vividus-plugin-ssh/src/test/java/org/vividus/ssh/exec/SshExecutorTests.java [hunk 14]
```diff
@@ -75,9 +75,9 @@ void testExecuteCommandsViaSshSuccessfully() throws Exception
...
         String commands = "ssh-command";
-        SshOutput sshOutput = sshExecutor.executeCommand(serverConfiguration, new Commands(commands), channel);
+        SshOutput sshOutput = sshExecutor.executeCommand(sshConnectionParameters, new Commands(commands), channel);
```
Comment (ikalinin1) [inline vividus-plugin-ssh/src/test/java/org/vividus/ssh/exec/SshExecutorTests.java:80]:
```suggestion
        var sshOutput = sshExecutor.executeCommand(sshConnectionParameters, new Commands(commands), channel);
```

File: vividus-plugin-ssh/src/test/java/org/vividus/ssh/exec/SshShellExecutorTests.java [hunk 15]
```diff
@@ -46,10 +46,10 @@ void shouldReturnExecChannelType()
     void shouldConfigureChannel()
     {
         ChannelShell channel = mock(ChannelShell.class);
-        ServerConfiguration serverConfiguration = new ServerConfiguration();
-        serverConfiguration.setAgentForwarding(true);
-        serverConfiguration.setPseudoTerminalEnabled(true);
-        sshShellExecutor.configureChannel(channel, serverConfiguration);
+        SshConnectionParameters sshConnectionParameters = new SshConnectionParameters();
```
Comment (ikalinin1) [inline vividus-plugin-ssh/src/test/java/org/vividus/ssh/exec/SshShellExecutorTests.java:49]:
```suggestion
        var sshConnectionParameters = new SshConnectionParameters();
```

File: vividus-plugin-ssh/src/test/java/org/vividus/ssh/sftp/SftpExecutionManagerTests.java [hunk 16]
```diff
@@ -24,21 +24,21 @@
...
+import org.vividus.ssh.SshConnectionParameters;
 
 class SftpExecutionManagerTests
 {
     @Test
     void shouldRunExecution() throws CommandExecutionException
     {
         SftpExecutor executor = mock(SftpExecutor.class);
-        ServerConfiguration serverConfiguration = new ServerConfiguration();
+        SshConnectionParameters sshConnectionParameters = new SshConnectionParameters();
```
Comment (ikalinin1) [inline vividus-plugin-ssh/src/test/java/org/vividus/ssh/sftp/SftpExecutionManagerTests.java:35]:
```suggestion
        var sshConnectionParameters = new SshConnectionParameters();
```

File: vividus-plugin-ssh/src/test/java/org/vividus/ssh/sftp/SftpExecutorTests.java [hunk 17]
```diff
@@ -57,15 +57,15 @@ void testExecuteCommandsViaSftpSuccessfully() throws Exception
         ChannelSftp channel = mock(ChannelSftp.class);
         String pwd = "/Users";
         when(channel.pwd()).thenReturn(pwd).thenReturn(pwd);
-        ServerConfiguration serverConfiguration = new ServerConfiguration();
-        serverConfiguration.setAgentForwarding(true);
-        serverConfiguration.setPseudoTerminalEnabled(true);
+        SshConnectionParameters sshConnectionParameters = new SshConnectionParameters();
```
Comment (ikalinin1) [inline vividus-plugin-ssh/src/test/java/org/vividus/ssh/sftp/SftpExecutorTests.java:60]:
```suggestion
        var sshConnectionParameters = new SshConnectionParameters();
```

File: vividus-plugin-ssh/src/test/java/org/vividus/ssh/sftp/SftpExecutorTests.java [hunk 18]
```diff
@@ -57,15 +57,15 @@ void testExecuteCommandsViaSftpSuccessfully() throws Exception
...
+        sshConnectionParameters.setAgentForwarding(true);
+        sshConnectionParameters.setPseudoTerminalEnabled(true);
         Commands commands = new Commands("pwd; cd ~; pwd");
-        SftpOutput sftpOutput = sftpExecutor.executeCommand(serverConfiguration, commands, channel);
+        SftpOutput sftpOutput = sftpExecutor.executeCommand(sshConnectionParameters, commands, channel);
```
Comment (ikalinin1) [inline vividus-plugin-ssh/src/test/java/org/vividus/ssh/sftp/SftpExecutorTests.java:64]:
```suggestion
        var sftpOutput = sftpExecutor.executeCommand(sshConnectionParameters, commands, channel);
```

File: vividus-plugin-ssh/src/test/java/org/vividus/ssh/sftp/SftpExecutorTests.java [hunk 19]
```diff
@@ -80,14 +80,14 @@ void testPopulateErrorStreamOnSftpError() throws Exception
         ChannelSftp channel = mock(ChannelSftp.class);
         SftpException sftpException = new SftpException(0, "error");
         doThrow(sftpException).when(channel).pwd();
-        ServerConfiguration serverConfiguration = new ServerConfiguration();
-        serverConfiguration.setAgentForwarding(true);
-        serverConfiguration.setPseudoTerminalEnabled(true);
-        SftpOutput sftpOutput = sftpExecutor.executeCommand(serverConfiguration, new Commands("pwd"), channel);
+        SshConnectionParameters sshConnectionParameters = new SshConnectionParameters();
```
Comment (ikalinin1) [inline vividus-plugin-ssh/src/test/java/org/vividus/ssh/sftp/SftpExecutorTests.java:83]:
```suggestion
        var sshConnectionParameters = new SshConnectionParameters();
```

File: vividus-plugin-ssh/src/test/java/org/vividus/steps/ssh/SshStepsTests.java [hunk 20]
```diff
@@ -53,96 +58,154 @@
...
+    void shouldConfigureDynamicConnection()
+    {
+        var connectionParametersTable = new ExamplesTable("|username |host       |port|password|\n"
+                        + "|admin    |10.10.10.10|22  |Pa$$w0rd|");
+        var steps = new SshSteps(null, variableContext, commandExecutionManagers, sshTestContext);
+        var key = "new-connection";
+        steps.configureSshConnection(key, connectionParametersTable);
+        var sshConnectionParametersArgumentCaptor = ArgumentCaptor.forClass(SshConnectionParameters.class);
+        verify(sshTestContext).addDynamicServerConfiguration(eq(key), sshConnectionParametersArgumentCaptor.capture());
+        SshConnectionParameters parameters = sshConnectionParametersArgumentCaptor.getValue();
```
Comment (ikalinin1) [inline vividus-plugin-ssh/src/test/java/org/vividus/steps/ssh/SshStepsTests.java:78]:
```suggestion
        var parameters = sshConnectionParametersArgumentCaptor.getValue();
```

File: vividus-plugin-ssh/src/test/java/org/vividus/steps/ssh/SshStepsTests.java [hunk 21]
```diff
@@ -53,96 +58,154 @@
...
         testPutFile(SftpCommand.PUT_FROM_FILE, filePath,
-            () -> sshSteps.copyFileOverSftp(filePath, DESTINATION_PATH, SERVER));
+            steps -> steps.copyFileOverSftp(filePath, DESTINATION_PATH, SERVER));
     }
 
-    private void testPutFile(SftpCommand command, String parameter, StepRunner stepExecutor)
+    private void testSshCommandsExecution(Map<String, SshConnectionParameters> staticConnections, String connectionKey)
             throws CommandExecutionException
+    {
+        CommandExecutionManager<SshOutput> executionManager = mockGettingOfCommandExecutionManager(Protocol.SSH);
```
Comment (ikalinin1) [inline vividus-plugin-ssh/src/test/java/org/vividus/steps/ssh/SshStepsTests.java:160]:
```suggestion
       var executionManager = mockGettingOfCommandExecutionManager(Protocol.SSH);
```

File: vividus-plugin-ssh/src/main/resources/spring.xml (hunk 4 above)
Comment (valfirst) [reply vividus-plugin-ssh/src/main/resources/spring.xml:14]: general change in order to keep everything under `org.vividus.ssh` package

**Review by ikalinin1 — APPROVED** [review]

**Review by uarlouski — APPROVED** [review]
